extensions/ql-vscode/src/skeleton-query-wizard.ts

```diff
@@ -0,0 +1,247 @@
+import { join } from "path";
+import { CancellationToken, Uri, workspace, window as Window } from "vscode";
+import { CodeQLCliServer } from "./cli";
+import { OutputChannelLogger } from "./common";
+import { Credentials } from "./common/authentication";
+import { QueryLanguage } from "./common/query-language";
+import { askForLanguage, isFolderAlreadyInWorkspace } from "./helpers";
+import { getErrorMessage } from "./pure/helpers-pure";
+import { QlPackGenerator } from "./qlpack-generator";
+import { DatabaseManager } from "./local-databases";
+import * as databaseFetcher from "./databaseFetcher";
+import { ProgressCallback } from "./progress";
+
+type QueryLanguagesToDatabaseMap = Record<string, string>;
+
+export const QUERY_LANGUAGE_TO_DATABASE_REPO: QueryLanguagesToDatabaseMap = {
+  cpp: "protocolbuffers/protobuf",
+  csharp: "dotnet/efcore",
+  go: "evanw/esbuild",
+  java: "google/guava",
+  javascript: "facebook/react",
+  python: "pallets/flask",
+  ruby: "rails/rails",
+  swift: "Alamofire/Alamofire",
+};
+
+export class SkeletonQueryWizard {
+  private language: string | undefined;
+  private fileName = "example.ql";
+  private storagePath: string | undefined;
+
+  constructor(
+    private readonly cliServer: CodeQLCliServer,
+    private readonly progress: ProgressCallback,
+    private readonly credentials: Credentials | undefined,
+    private readonly extLogger: OutputChannelLogger,
+    private readonly databaseManager: DatabaseManager,
+    private readonly token: CancellationToken,
+  ) {}
+
+  private get folderName() {
+    return `codeql-custom-queries-${this.language}`;
+  }
+
+  public async execute() {
+    // show quick pick to choose language
+    this.language = await this.chooseLanguage();
+    if (!this.language) {
+      return;
+    }
+
+    this.storagePath = this.getFirstStoragePath();
+
+    const skeletonPackAlreadyExists = isFolderAlreadyInWorkspace(
+      this.folderName,
+    );
+
+    if (skeletonPackAlreadyExists) {
+      // just create a new example query file in skeleton QL pack
+      await this.createExampleFile();
+      // select existing database for language
+      await this.selectExistingDatabase();
+    } else {
+      // generate a new skeleton QL pack with query file
+      await this.createQlPack();
+      // download database based on language and select it
+      await this.downloadDatabase();
```

Comment:
It might be helpful to send some messages to the extension logger before/after each step, to help us (and users) with debugging.

Reply:
Do we do this anywhere else? Just looking for examples.